crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -797,6 +803,25 @@ impl<'db> TypeInferenceBuilder<'db> {
         self.types.definitions.insert(definition, context_expr_ty);
     }
 
+    fn infer_except_handler_definition(
```

Comment:
It looks to me like we will still have double-inference of some except handlers in this version of the PR, because now we are entirely missing the code that we need to add in `infer_try_statement`, which needs to check, for each handler, whether it is one that would create a definition (that is, it has a `name`), and if so, call `self.infer_definition(handler.name)` (this is where it matters which node is used as the DefinitionNodeKey; that's what we have to look up the definition by), which will use the per-definition inference query to infer that handler and merge the types back into the scope-level inference.

Otherwise, with the current code in this PR, scope-level inference and definition-level inference will separately double-infer the same except handler (and thus, in the future, potentially double-emit diagnostics.)

Fixing https://github.com/astral-sh/ruff/issues/13168 will allow us to catch this automatically, instead of having to watch for it in code review.

Reply:
I _think_ this should be fixed now after my latest pushes!

Reply:
Yes, looks good!